RFD 127: Encrypted Session Recording

eriktate opened the pull request:
This PR proposes an RFD for encrypting session recordings using `age` encryption.

eriktate committed:
adding encrypted session recording RFD

zmb3 commented on rfd/0207-encrypted-session-recordings.md:
```diff
@@ -0,0 +1,284 @@
+Admins are free to configure their own public key recipients provided they are
+generated by a supported algorithm.
+
+As a convenience, we will also add a new `keygen` utility to `tsh` for key
```
Would this be more appropriate for `tctl` than `tsh`? It feels like an administrative workflow to me, not an end-user workflow.

eriktate replied:
Yeah I think that makes more sense. At one point it was a `tctl` command, but the similarity to `ssh-keygen` and the fact that `tsh` _has_ to integrate with HSMs in order to playback made me wonder if it would be worth it to just do the entire workflow with `tsh`. `Administrative workflows running through `tctl` seems like a more compelling reason, though

zmb3 commented on rfd/0207-encrypted-session-recordings.md:
```diff
@@ -0,0 +1,284 @@
+
+### Decryption and Replay
+
+Replay of encrypted session recordings will happen through `tsh play`. For
```
What will we do about desktop session recordings, which can only be played via the UI?

eriktate replied:
With the current design they wouldn't be replayable. In theory we could download and decrypt all of the PNGs with a command, but we would have to stitch the chunks together into full PNGs and the UX would be pretty poor.

I had an idea to add something like a `tsh decrypt-server` to support replays on web. It would spin up a local server capable of unwrapping the decryption key at an address that could be pasted into the web UI. The web UI would send encrypted recording data in and get decrypted recording data out, all local to the user's machine. I've seen a similar setup work before, but I haven't fully fleshed it out yet.

zmb3 replied:
It's definitely an option - I have always thought of doing a `teleport play desktop` or `tsh play desktop` that would spin up a local webserver that can play a desktop session for offline playback, though this might be a big effort.

Can auth decrypt the recordings on the fly? Recordings play through auth's `StreamSessionEvents` API, so if auth has access to the HSM can it stream decrypted events to the client?

zmb3 commented on rfd/0207-encrypted-session-recordings.md:
```diff
@@ -0,0 +1,284 @@
+convenience, it should be possible to provide the identity used for decryption
+through stdin or with a file path. This would add the flags `--identity-stdin` and
+`--identity-path` respectively. When playing recordings managed by an HSM, the
+identity will be wrapped and included as the first line of the recording which will
```
What do you mean by "first line" here? Session recordings are an ordered, gzipped stream of proto events. How does the concept of a line come into play here?

Also, IIRC our recordings have some sort of header with a version and maybe other metadata. Do we plan to update this in a backwards compatible way?

zmb3 commented on rfd/0207-encrypted-session-recordings.md:
````diff
@@ -0,0 +1,284 @@
+
+### Teleport admin replaying encrypted session with a given private key
+```bash
+cat ~/my-key | tsh play --identity-stdin 49608fad-7fe3-44a7-b3b5-fab0e0bd34d1
````
Nit: `identity-stdin` sounds like _user identity_, not like something necessary for decryption.

eriktate replied:
Maybe `--private-key-stdin` instead? The "identity" was just in keeping with how `age` refers to things, but in retrospect I agree it's way less obvious

zmb3 replied:
Sure, or `--decryption-key-stdin` to be even more clear.

zmb3 commented on rfd/0207-encrypted-session-recordings.md:
````diff
@@ -0,0 +1,284 @@
+
+### Teleport admin replaying an encrypted session recording containing an AWS CloudHSM wrapped key
+```bash
+tsh play --hsm=aws cead9efc-f0d0-4ff3-ad34-1346adc673e8
````
What else would be necessary here? Is the `aws` CLI required?

eriktate replied:
Since this is now happening in the auth service, and we already do this for CA keystores, I'll probably just leverage that instead of introducing new configuration for the same thing

zmb3 commented on rfd/0207-encrypted-session-recordings.md:
```diff
@@ -0,0 +1,284 @@
+- Sessions can be recorded when there are recipients configured
+- Encrypted sessions can be played back by passing any recipient's identity to `tsh play`.
+- `tsh keygen --hsm=aws` provides `age_x25519` public key with a wrapped private key (can be verified by comparing key length with standard `tsh keygen`)
+- Sessions encrypted using an HSM wrapped key can be played back using the same `--hsm` configuration.
```
Should we make a change and backport it to old versions of `tsh` that can at least detect that it's an encrypted recording and advise the user to upgrade before they can play it?

nklaassen commented on rfd/0207-encrypted-session-recordings.md:
```diff
@@ -0,0 +1,293 @@
+
+This document should fulfill the following requirements:
+- Ability to encrypt session recording data during intermediate disk writes and long term storage.
+- Ability to provide an externally managed public key for encryption.
```
is this a direct customer ask? I'm wondering why we don't want to use a Teleport-managed key so that sessions could be decrypted on Auth and played as normal in the UI and from tsh. Or if that could be an option, with the externally-managed key being opt-in only for customers who explicitly want that

eriktate replied:
I dropped this entirely for now. We can always reintroduce this if we have an actual need for it in the future

ikkisoft commented on rfd/0207-encrypted-session-recordings.md:
```diff
@@ -0,0 +1,293 @@
+set of `--hsm*` parameters. More information about this in
+[HSM Configuration](#hsm-configuration).
+
+### Session Recording Modes
```
Not sure what are the clients requirement for this, but it seems over-complicated to me. I would pick one mode only if this is not driven by client requests

eriktate replied:
This is actually just a listing of recording modes we already support today. I suppose we could only enable encryption for a limited set of modes, but I don't think it would save us much complexity unless we only supported `sync` modes.

ikkisoft commented on rfd/0207-encrypted-session-recordings.md:
```diff
@@ -0,0 +1,293 @@
+using identities associated with any recipient used during encryption, it is possible
+to verify that new recordings are still retrievable before removing the rotated
+recipient. Reencrypting recordings captured before the rotation is not in scope for
+this implementation. A potential method for ensuring continued access to historical
```
A possible option here could be to encrypt all recording with a long-term key associated to the first admin user created on Teleport. Having said that, if key rotation is not required for old recordings, I wouldn't increase the complexity of keys management

eriktate replied:
I agree, and I don't think this is a requirement we need to support :+1:

eriktate commented:
Important changes in the latest commit:
- Dropped manual key management since it isn't explicitly required right now
- Moved decryption to the auth service
- Added a CA for easier keystore integration, including HSMs
- Data encryption scheme is now just the standard envelope encryption provided by `age`

The reason for wrapping a separate `age` private key is to enable the proxy service and nodes to use the public key directly without having access to the backing keystore.

eriktate commented on rfd/0127-encrypted-session-recordings.md:
```diff
@@ -0,0 +1,169 @@
+In order to generate keys, a new session encryption CA will be added. The CA
+will be configured through the existing `auth_service.ca_key_params` block.
+A new `Ed25519` keypair will be generated and added to the CA's active keyset.
+For keystores that support wrapping, including HSMs, the private key will be
+wrapped before being added. The public key will be used by `age` to wrap data
+encryption keys at the point of encryption, and the private key will be used
+within the auth service to decrypt during playback.
```
> Either that or we use the HSM/KMS to wrap a software Ed25519 key which we store encrypted in the CA object, and use that as the Identity and Recipient for age. Maybe that's what you're getting at here?

Yeah this is exactly what I was going for. I wasn't sure how realistic it would be to implement the `Recipient` half of the `age` plugin in a way that would be compatible with the HSM/KMS native decryption. That's the only reason for the intermediate `Ed25519` keypair, so if it's reasonable for us implement directly against the HSM/KMS managed keys I'm all for that.

As far as the keys currently involved in the RFD:
- The `Ed25519` keypair would be generated by Teleport to be used with `age` as the `Identity` and `Recipient`. This doesn't necessarily have to be `Ed25519`, I just chose it due to Teleport and `age` both being compatible.
- The keys required to wrap/unwrap the `Identity` with the HSM/KMS. If this were implemented using the native encrypt/decrypt functions then I think it could just be a symmetric key fully managed by the HSM.
- The wrapped data encryption keys that `age` generates and packages during file encryption.

eriktate commented on rfd/0127-encrypted-session-recordings.md:
```diff
@@ -0,0 +1,169 @@
+
+In order to generate keys, a new session encryption CA will be added. The CA
+will be configured through the existing `auth_service.ca_key_params` block.
+A new `Ed25519` keypair will be generated and added to the CA's active keyset.
```
We can definitely use a different algorithm here, but this keypair would be encrypted and stored in the CA object like you mentioned in the comment further down.

eriktate commented on rfd/0127-encrypted-session-recordings.md:
```diff
@@ -0,0 +1,325 @@
+
+  // RotateSessionRecordingKeys rotates the keys associated with encrypting and
+  // decrypting session recordings.
+  rpc RotateSessionRecordingKeys(RotateSessionRecordingKeysRequest) returns (RotateSessionRecordingKeysResponse);
```
I went back and forth on whether or not this was the right place for this. I'm not sure it warrants a new service, but I'm certainly open to feedback here.

nklaassen commented on rfd/0127-encrypted-session-recordings.md:
I'm looking for some mention of how we're going to generate these wrapping keys in the HSM and/or KMS. Do we need to use any new AWS KMS APIs, what about GCP? do we need any new PKCS11 calls for decryption, do most/all HSMs support those with RSA keys?

nklaassen commented on rfd/0127-encrypted-session-recordings.md:
```diff
@@ -0,0 +1,326 @@
+keys are included with their rotated private keys in order to facilitate faster
+searching.
+
+### Key Rotation
```
will we ever delete keys from the HSM/KMS? how will we prevent these keys from being deleted by the existing DeleteUnusedKeys function?

eriktate replied:
I think the easiest way would be to fetch the active keys from the `SessionRecordingConfigV2` resource and include them in the list of `allKeysInUse`. That should keep them around and also clean them up automatically when they're rotated out.

eriktate replied:
Good call 🤔 I think we could just use a different label for these keys. That should keep older auth servers from accidentally deleting them and newer auth servers can target keys with both expected labels

nklaassen commented on rfd/0127-encrypted-session-recordings.md:
```diff
@@ -0,0 +1,326 @@
+For the rest of this section I will refer to the public key as the `Recipient`
+and the private key as the `Identity` in keeping with `age` terminology.
+
+The auth server generating the keys will then use the configured CA keystore to
+generate an `RSA` keypair used to encrypt, or wrap, the `Identity`. Once
```
will you use OAEP encryption? i'd like to see somewhere a mention of the encryption algorithm and any parameters and how we will do decryption with PKCS11, AWS KMS, GCP KMS

eriktate commented on rfd/0127-encrypted-session-recordings.md:
```diff
@@ -0,0 +1,326 @@
+the auth service and can be fulfilled by any auth server with an active wrapped
+key.
+
+When an auth server receives a rotation request for wrapped keys, it will:
```
It looks we have one instance of this with `tctl alerts ack` and `tctl alerts ack ls`, but I'm not sure if that's actually working based on the help text. Maybe something like `auth keys rotate` would be better?

eriktate commented on rfd/0127-encrypted-session-recordings.md:
```diff
@@ -0,0 +1,326 @@
+  symmetric encryption of the recording data.
+- `Identity` and `Recipient` keys used by `age` to wrap and unwrap file keys.
+  These are an asymmetric `X25519` keypair.
+- HSM/KMS-owned `RSA` keypairs used to wrap the `Identity` key. These allow the
+  HSM or KMS keystore to "own" decryption without requiring proxy or host nodes
+  to make encryption requests to the auth server.
```
This is mostly just to point out that the HSM is required for decryption, but it doesn't need to be involved with encryption of `age` file keys. I can be more straightforward about that.

Although I guess this would technically still be true if we wrapped them using OEAP against the public RSA key. I think we'd lose the property of making rotated keys accessible to new auth servers, though

nklaassen commented on rfd/0127-encrypted-session-recordings.md:
````diff
@@ -0,0 +1,326 @@
+auth servers without immediate access to an active key will return an error to
+the client letting them know the rotation should be tried again.
+
+#### `tctl` Changes
+
+Key rotation will be handled through `tctl` using a new subcommand:
+```bash
+tctl auth rotate recordings --type=data # for rotating X25519 pairs
+tctl auth rotate recordings --type=wrap # for rotating HSM/KMS backed wrapping keys
+```
````
(i think you replied on the wrong thread but I'll reply here)

> It looks we have one instance of this with `tctl alerts ack` and `tctl alerts ack ls`, but I'm not sure if that's actually working based on the help text. Maybe something like auth keys rotate would be better?

That is some kind of hack where `tctl alerts ack ls` is not a real subcommand, it accepts an argument for the alert id, and if the arg happens to be "ls" then it takes over https://github.com/gravitational/teleport/blob/master/tool/tctl/common/alert_command.go#L144-L146

You'd have to do some similar hack with an argument to make `tctl auth rotate recordings` works but then yeah the help text would be weird, let's try to avoid that.

Maybe something under `tctl recordings`? Idk like `tctl recordings encryption rotate`

eriktate replied:
I think `tctl recordings encryption rotate` would work fine too. I would only wonder if being a subcommand of `auth` makes it more discoverable since `auth` already handles CA rotation?

zmb3 replied:
+1 for putting it under `tctl recordings`.

eriktate commented on rfd/0127-encrypted-session-recordings.md:
```diff
@@ -0,0 +1,373 @@
+
+### Encryption Algorithm
+
+This RFD assumes the usage of [age](https://github.com/FiloSottile/age), which
+was chosen for its provenance, simplicity, and focus on strong cryptography
+defaults without requiring customization. The formal spec can be found
+[here](https://age-encryption.org/v1). Officially supported key algorithms are
+limited to X25519 (recommended by the spec), Ed25519, and RSA. Support for
+other algorithms would either have to be requested from the upstream or
+manually implemented as a custom plugin.
```
We could definitely consider alternatives if FIPS should be part of the initial requirements. If we don't need it at the start, though, maybe we could ask about a potential FIPS mode with the `age` maintainers?

To be fair, though, we aren't currently making much use of the multi-recipient property of `age` so a FIPS compliant alternative likely wouldn't cause too much change in the overall implementation

eriktate commented on rfd/0127-encrypted-session-recordings.md:
```diff
@@ -0,0 +1,373 @@
+implementing different key algorithms only affect the crypto involved with
+wrapping and unwrapping data encryption keys.
+
+In both proxy and node recording modes, the public `X25519` key used for
+wrapping `age` data keys will come from
+`session_recording_config.status.active_keys`. All unique public keys present
+will be added as recipients.
```
I did consider generating recipients per auth server, because it makes a lot of things simpler. We could use the `RSA` public key directly to wrap `age` file keys and remove the intermediate `X25519` keypair (and all of the extra management) entirely.  There would also only be one key rotation scenario to consider. Maintaining access to historical encrypted recordings becomes harder, though. If a newly added auth server fails to decrypt because it doesn't have an identity associated with any of the recording's recipients, it would need some way to route that request to an auth server that could. Which also means old recordings could effectively be lost if all auth servers with access to the original keys were replaced.

zmb3 commented on rfd/0127-encrypted-session-recordings.md:
```diff
@@ -0,0 +1,373 @@
+
+For the most part, the user experience of encrypted session recordings is
+identical to non-encrypted session recordings. The only notable change is the
+addition of the `tctl auth rotate recordings` subcommand for rotating keys
```
Is it `tctl auth rotate recordings` or `tctl auth keys rotate`?

eriktate replied:
It should be `tctl auth keys rotate` for now, this and the example commands should be fixed now :+1:

zmb3 commented on rfd/0127-encrypted-session-recordings.md:
````diff
@@ -0,0 +1,373 @@
+
+### Teleport admin replaying encrypted session using tsh
+```bash
+tsh play 49608fad-7fe3-44a7-b3b5-fab0e0bd34d1
````
What about playing from file? Will `tsh play 49608fad-7fe3-44a7-b3b5-fab0e0bd34d1.tar` print an error for encrypted recording files?

espadolini commented on rfd/0127-encrypted-session-recordings.md:
```diff
@@ -0,0 +1,380 @@
+strategy, but whether or not they are handled async or sync has different
+requirements.
+
+In sync modes the session recording data is written immediately to long term
+storage without intermediate disk writes. This means we can simply instantiate
+an age encryptor at the start of the session and encrypt the recording data as
+it's sent to long term storage.
+
+In async modes the session recording data is written to intermediate `.part`
+files. These files are collected until they're ready for upload and are then
+combined into a single `.tar` file. In order to encrypt individual parts, we
+will build a special `io.Writer` that contains a single instance of the age
+encryptor that can proxy writes across multiple files. This will require
+maintaining a concurrency-safe mapping of in-progress uploads to encrypted
+writers which incurs a bit of added complexity. However it intentionally avoids
+any sort of intermediate key management which seems a worthwhile tradeoff.
```
How does this interact with the chunked uploading of parts and the potential for recordings to be terminated while incomplete? AIUI the age format assumes that the file is fully available and must be terminated correctly for the data to be safely decrypted.

eriktate replied:
Edoardo and I talked about this in Slack, but I think it would be possible to to encrypt each chunk separately, allow them to be combined in long term storage, and then split them based on the `age` header at decryption time. We would take a performance hit and a bit of a storage hit, though. I'm still trying to work out if there's a way around that.

eriktate commented on rfd/0127-encrypted-session-recordings.md:
```diff
@@ -0,0 +1,380 @@
+  // ActiveKeys is a list of active, wrapped X25519 private keys. There should
+  // be at most one wrapped key per auth server using the
+  // SessionRecordingConfigV2 resource unless keys are being rotated.
+  repeated WrappedKey ActiveKeys = 2 [
```
Because I deleted a top level `PublicKey` field that used to be 1 :sweat_smile: Should be fixed

eriktate commented on rfd/0127-encrypted-session-recordings.md:
````diff
@@ -0,0 +1,441 @@
+
+Key rotation will be handled through `tctl` using a new subcommand:
+```bash
+tctl recordings encryption rotate
````
I'm not sure if it needs to be. It probably could be if we wanted to block until rotation completes, but there aren't really any phases to navigate other than "rotation in progress" and "rotation complete"

eriktate commented on rfd/assets/0127-age-high-level.png:
It's mostly just to show that there can be multiple stanzas and what a `wrapped filekey` actually is. If it's not super clear I can probably just reduce it down to a single stanza and add a note that there can be multiple

nklaassen commented on rfd/0127-encrypted-session-recordings.md:
```diff
@@ -0,0 +1,489 @@
+
+// EncryptionKeyPair is a keypair used for encrypting and decrypting data.
+message EncryptionKeyPair {
+  // PublicKey is the SSH public key.
+  bytes PublicKey = 1 [(gogoproto.jsontag) = "public_key,omitempty"];
+  // PrivateKey is the SSH private key.
+  bytes PrivateKey = 2 [(gogoproto.jsontag) = "private_key,omitempty"];
```
the mention of SSH here is a bit of a red herring. Will these be serialized in OpenSSH format, except for HSM keys, which will follow the existing HSM key format I presume?

eriktate replied:
I accidentally left `SSH` in these comments from the previous proto definition. In practice these will either be `X25519` keys stored as `Raw`, or HSM/KMS keys which will follow the existing key format

nklaassen commented on rfd/0127-encrypted-session-recordings.md:
```diff
@@ -0,0 +1,489 @@
+with an `age` encryption writer.
+
+In async modes the session recording data is written to intermediate `.part`
+files. These files are collected until they're ready for upload and are then
+combined into a single `.tar` file. In order to encrypt individual parts, we
```
is this what already happens, all parts being into a single .tar for upload? if it's new i wonder if it could be a problem if the .tar is quite big, could take a lot of memory

eriktate replied:
It is but the `.tar` is parsed into individual events that are sent to the auth service in more or less the same way `sync` modes work. Since the `.tar` is already encrypted that won't be possible, but we can still chunk the upload to the auth service.

nklaassen commented on rfd/0127-encrypted-session-recordings.md:
```diff
@@ -0,0 +1,489 @@
+In async modes the session recording data is written to intermediate `.part`
+files. These files are collected until they're ready for upload and are then
+combined into a single `.tar` file. In order to encrypt individual parts, we
+will build a special `io.Writer` that contains a single instance of the age
+encryptor that can proxy writes across multiple files. This will require
+maintaining a concurrency-safe mapping of in-progress uploads to encrypted
+writers which incurs a bit of added complexity. However it intentionally avoids
+any sort of intermediate key management which seems a worthwhile tradeoff.
```
what will happen if the teleport agent dies/restarts mid-session, can it restore the in-progress recording?

nklaassen commented on rfd/0127-encrypted-session-recordings.md:
```diff
@@ -0,0 +1,489 @@
+into events to be uploaded to the auth service is not possible. Instead the
+auth service must accept a binary upload of the `.tar` file which it can
+then proxy to long term storage. This will be done using the
+`UploadEncryptedSessionRecording` streaming RPC. When using S3 compatible long
+term storage, the auth service will need to use the `PutObject` API rather than
+performing a multipart upload. This is to prevent any loss of parts that would
+make decryption impossible.
```
i assume there's a reason we currently do multi-part uploads, could sessions be too large for a single PutObject? just a question idk

eriktate replied:
The maximum size for `PutObject` is 5GB, so I wouldn't think it's super likely? As far as I can tell we do the multi-part upload so that we aren't holding onto large amounts of recording data in memory while we wait for the session to conclude. The issue of the agent crashing halfway through definitely needs to be addressed, though. I think it would force us to encrypt each part/chunk individually similar to how `sync` modes would work. Even so, we would need this new RPC to upload the encrypted chunks since we can't deconstruct them back into events at the node/proxy level.

nklaassen commented on rfd/0127-encrypted-session-recordings.md:
```diff
@@ -0,0 +1,489 @@
+  `Decrypt` action of the KMS API.
+- GCP CKM supports generating keys of type `RSAES_OAEP_2048_SHA256` which can
+  be used when calling the `AsymmetricDecrypt` API from their go SDK. Worth
+  noting that keys generated for signing can not be used for decryption, so key
+  generation with a GCP backend will need to be modified slightly.
```
i think this is the case for AWS too

espadolini commented on rfd/0127-encrypted-session-recordings.md:
```diff
@@ -0,0 +1,490 @@
+
+message UploadEncryptedSessionRecordingResponse {}
+
+service AuthService {
```
New RPCs should be in separate services unless there's a good reason to keep them in the giant `proto.AuthService`.

espadolini commented on rfd/0127-encrypted-session-recordings.md:
```diff
@@ -0,0 +1,490 @@
+message SessionRecordingConfigSpecV2 {
+  // existing fields omitted
+
+  SessionRecordingEncryptionConfig Encryption = 3 [
+    (gogoproto.jsontag) = "encryption"],
+    (gogoproto.nullable) = true,
+  ]
```
That being said, this is probably an important enough bit of configuration (that we might not necessarily allow in cloud outside of select tenants) to deserve its own separate configuration singleton (which won't need all the gogoproto cruft, added bonus).

espadolini replied:
Yes, especially because - as I just realized - the data format as written would imply that every agent in the cluster will sync the full list of keys, including the historical ones (and the cache fanout of the auth will have to deal with dropping secrets from it for each of them).

Considering that this is data that will be rarely updated (so transactional writes are not a concern) and that will be synced to the whole cluster (!) while needing processing (!!) because it contains private key material (!!!) I'd ask you to consider changing the storage format to have a dedicated item for the currently intended wrapped public keys that the nodes in the cluster can cache, with other items dedicated for auth-only consumption with the encrypted private key material and all the historical keys. (cc @fspmarshall)

p.s. what do you think of naming the two layers of keys "Recording Encryption Key" and "Key Encryption Key" instead of referring them as wrapped and wrapping?

eriktate replied:
All great points! I think that's partially in line with what I was proposing in the other thread about resource size. I hadn't considered the fact that the rest of the cluster shouldn't have access to the private half of the wrapped keys, so I'll take a look a stab at rethinking how this gets stored

eriktate replied:
> p.s. what do you think of naming the two layers of keys "Recording Encryption Key" and "Key Encryption Key" instead of referring them as wrapped and wrapping?

I went ahead and did this while adjusting the protos because I think this makes more sense. I wasn't sure how specific to session recording I should keep the naming, but now that wrapped keys are managed in a `recording_encryption` I think `recording_encryption_key` and `key_encryption_key` are a lot easier to use in a sentence lol

espadolini commented on rfd/0127-encrypted-session-recordings.md:
```diff
@@ -0,0 +1,490 @@
+}
+// SessionRecordingConfigStatusV2 contains all of the current and rotated keys
+// used for encrypted session recording.
+message SessionRecordingConfigStatusV2 {
+  EncryptionKeySet KeySet = 1 [(gogoproto.jsontag) = "keyset"]
+}
```
This is stored in the backend, likely as a resource, probably in json, each wrapped key is going to be something like 2KiB and we only have like 400KiB to play with in a single backend item - using a single item has the potential to be _very_ problematic after a cluster has existed for a while.

eriktate replied:
Do we have any existing features that split across arbitrary numbers of backend items? Since we can't really throw out rotated keys, I agree this will definitely be a problem for someone at some point

eriktate replied:
Since I split this into new resources, the base path would now be `/recording_encryption/rotated_keys/`. I'd probably just take the hex encoded sha256 of the public key as the fingerprint.

espadolini commented on rfd/0127-encrypted-session-recordings.md:
```diff
@@ -0,0 +1,490 @@
+  `true`. This is done in place of removing all other keys in order to avoid
+  delays or failures that might come from other auth servers not having
+  immediate access to an active `WrappedKey`.
+- Replace its own active `WrappedKey` with the newly created one, moving the
+  original value to the rotated keys list.
```
Doing the rotation in phases also has the benefit of disallowing accidental concurrent double rotations (because initiating a rotation is a separate step from finalizing a rotation), which have the potential to be incredibly confusing at best and data loss incidents otherwise.

eriktate replied:
Since all unique public keys present in the active list are used as recipients, I don't think there's a situation where rotation would block an auth server from encrypting or decrypting. The rotation would create a second active keypair which would be used with the original to encrypt recordings until all auth servers finished their rotation. The old key would be considered rotated out when it no longer exists as an active wrapped key for any auth server.

The double rotation is a good call out, though 🤔 I suppose we could prevent a rotation if any active key has the `rotate` flag? Not sure if that would be acceptable over having control over which phases run

eriktate replied:
I talked to Tim earlier today and we could potentially turn rotation into an interactive command that waits for all auth servers to provision new keys. Once they do, the admin would be prompted to continue which would shift all of the old keys out of the active list and into the rotated list. This would also give us an opportunity to let the admin know if something went wrong (e.g. an orphaned key that isn't being rotated) and allow them to make a decision about how to deal with that before carrying on with the rotation. If we went with this instead of a true phased approach, then we would block new rotation attempts while at least one active key is marked for rotation.

eriktate replied:
I typed most of a reply and realized I'm just describing rotation phases :sweat_smile:

I also realized that the `rotate` bool on wrapped keys should actually be a state enum with `active`, `rotating`, and `rotated` since there's no reliable way to know a key has already been rotated otherwise. Then the process would be an RPC to kick off rotation which marks all active keys as `rotating`. Each auth server would perform its own rotation, marking their original keys as `rotated` after. A second RPC could move all rotated keys out of the active set, but only if there are no active keys marked with `rotating`. 

The rotation phase would just be derived from the wrapped keys as:
- All keys `active`? No rotation in progress
- At least on key `rotating`? Rotation in progress
- All keys `active` or `rotated`? Rotation is finished and ready for second RPC

eriktate replied:
I'm honestly still not 100% convinced the second RPC is necessary, though. There's not really much risk involved with automatically shifting out the rotated keys as far as I can tell. We can prevent double rotations by returning an error if one's already in progress, and we could even store a timestamp to warn users if a rotation had already been performed recently. Making admins aware of abandoned keys is useful, but we could do that without requiring them to "finish" a rotation.

espadolini replied:
> There's not really much risk involved with automatically shifting out the rotated keys as far as I can tell.

Does the scenario I pictured not count?

> key KA might be usable by auths AX and AY but newly generated keys might not have the same property - for example when migrating from a software key to HSMs - so auth AX generates key KB and marks KA as rotated, auth AY might not have access to KB because of configuration changes, and all recordings made until it has generated a new key and the matching public key is propagated will forever be inaccessible to it

eriktate replied:
Oh I think I misunderstood this the first time I read it 🤔 I didn't realize this was possible, but the `backendForNewKeys` on the `keystore.Manager` suddenly makes much more sense. Yeah that would be problematic and software keys are going to be particularly vulnerable to this. For CA rotations, do we just rely on giving enough time? Or is there some other mechanism for figuring out progress?

espadolini replied:
I would argue that any sort of fully automatic CA rotation in Teleport should be discouraged and that human knowledge about any sort of KMS or hardware setup is required to be sure to not cause problems; the default automatic rotation uses 30 hours per phase (and there's always the possibility of manually rolling back until the rotation is finalized at the end), and the interactive CA rotation tool polls data from instance and resource heartbeats to figure out if a host CA rotation phase is "done". I'm not sure how we'd be able to automatically know if  all auths are ready to decrypt recordings with the new set of keys, since we can't really know which auths even exist at any given moment. 🥴

eriktate replied:
I updated the RFD to add additional `tctl` subcommands for handling rotations. I also mentioned emulating CA rotation delays before allowing a rotation to be completed.

@espadolini would you mind giving this one more pass when you have some time?

viktor-doyensec-teleport commented:
Hey team! I just finished my review of the RFD. Overall, I think the outlined design will provide a sufficient framework for a secure implementation of the screen recording encryption.

The only think I'd like to highlight, as also pointed out by @espadolini, is potential for race conditions during key rotation. Any issue that arise during the key update operations may result in an unexpected/inconsistent application state, and potentially lead to corrupted recordings. So ensure that all key change operations are performed as an atomic operation, to prevent any partial changes from being picket up by other auth services running on the cluster.

Let me know if you have any other questions!

eriktate commented on rfd/0127-encrypted-session-recordings.md:
```diff
@@ -0,0 +1,490 @@
+
+### Key Rotation
+
+Rotating keys used for session encryption requires rotating both the `X25519`
+keypair used with `age` and the keystore-backed wrapping keys. Because `age`
+allows us to encrypt with multiple recipients, all keys can be rotated in a
+single action.
```
This is an excellent point 😅 In the current model, I think we would have to re-encrypt all rotated keys accessible to the auth server at the same time that its active key is rotated out.

espadolini replied:
Do we need more than one wrapped key, then?

eriktate replied:
As far as active keys, there would be one per auth server. Rotated keys would remain wrapped as well, so the total number of wrapped keys is dependent on how many rotations have been done

espadolini replied:
What do you mean by "per auth server"? I thought we were going to have one if software keys were in use and one per distinct KMS/HSM if hardware/external keys were in use.

eriktate replied:
> do we need more than one active at a time?

It's mostly for the rotation case so that an auth server doesn't have to wait for its wrapped key to be fulfilled before it can serve replay requests. It would also come in handy during a rollback since all recordings captured during rotation will still be decryptable by the original REK.

> Who decides which KEK is supposed to reencrypt which REK at rotation time?

At rotation time, all active keys are marked as `rotating` and a new key is generated and wrapped by whichever auth server handles the rotation request. This will prompt all auth servers in the cluster to publish an unfulfilled key with their own KEK (similar to what will happen when a new auth server is added to the cluster) and those unfulfilled keys can only be fulfilled by REKs that aren't marked as `rotating`.

> And if all the REKs are encrypted by all the KEKs, wouldn't they just be equivalent?

I don't think I understand what you mean by equivalent here 🤔 They would be equivalent in practice for as long as they're both in the active set, but once the rotation completes and the old REK is rotated out, new recordings will no longer be decryptable by that key.

espadolini commented on rfd/0127-encrypted-session-recordings.md:
```diff
@@ -0,0 +1,634 @@
+data from being lost in the event of an agent crashing, each part will be
+encrypted individually much like the encrypted batches in `sync` modes. Because
+the data is already encrypted, exploding the final `.tar` file back into events
+to be uploaded to the auth service is not possible. Instead the auth service
+must accept a binary upload of the `.tar` file which it can then proxy to long
+term storage. This will be done using the `UploadEncryptedSessionRecording`
+streaming RPC. Each part will then be written to long term storage using ourt
+existing multipart upload functionality.
```
Will the auth server check each part to confirm that when decrypted it contains valid events in the correct format?

eriktate replied:
I haven't planned on doing this, but it could? I guess without this you could potentially upload arbitrary data provided it had a reasonable header 🤔

espadolini commented on rfd/0127-encrypted-session-recordings.md:
```diff
@@ -0,0 +1,634 @@
+the key ID. For HSMs integrated over PKCS#11, the auth server's host UUID is
+attached to the key ID and will be used to determine access.
+
+In order to avoid unintended automatic deletion, keys provisioned for encrypted
+session recording will be tagged with a different label where applicable. This
+will prevent older auth servers from deleting this new keytype and allow new
+auth servers to identify which keys are no longer in use before deletion.
```
Which keys are we referring to here? The key encryption keys or the recording encryption keys? It's not clear if the KEKs are going to be accumulated forever or if we're going to re-encrypt the REKs as their rotation happens (as well as encrypting new REKs as they get rotated in).

espadolini commented on rfd/0127-encrypted-session-recordings.md:
```diff
@@ -0,0 +1,634 @@
+  strategy, but whether or not they are handled async or sync has different
+  requirements.
+
+In sync modes the session recording data is written immediately to the auth
+service without intermediate disk writes. The auth service then handles
+batching and multipart upload of events to long term storage. In order to
+ensure resilience to partial uploads of parts, each batch will need to be
+encrypted individually on the auth service. This will likely mean wrapping the
+`gzipWriter` [used by slices](https://github.com/gravitational/teleport/blob/master/lib/events/stream.go#L669)
+with an `age` encryption writer.
```
Parts must be padded up to 5MiB (unless they're the last part) to support S3 multipart uploads, are we going to do that outside of the age envelope or can we do padding inside and get to exactly 5MiB on the outside? i.e. if we add one byte of plaintext, will we have one more byte of ciphertext?

eriktate replied:
The padding will be outside of the age envelope, so the file contents are:
```
24-byte header | encrypted, gzipped proto messages | padding up to 5MiB if required
```

eriktate replied:
Yeah definitely. I wasn't originally planning on changing the header format since we can inspect that the data segment starts with the `age` header, but after implementing that yesterday I think it might be better to add an extra byte for a bitflag that stores whether or not the data is encrypted. It looks like the header is intentionally 8 byte aligned, so I'd probably pad another 7 bytes of "reserved" space. I'll add this to the RFD

gl-mc commented:
The change request seems to use RSA 2048 bit keys, which are at the 112 bit security level. It would be better to use keys with 3072 bits or better, as they operate on the 128 bit security level, in line with the other cryptographic algorithms used (ChaCha20-Poly1305, X25519, SHA-256).

The key rotation seems to be overly complicated and involve re-keying of existing recordings. An alternative approach that does not require re-keying could provide benefits.

There does not seem to be a security benefit or performance benefit in wrapping X25519 keys with RSA keys. Technically, that is not possible in HSMs as they are both asymmetric key types. Recommended hosting the X25519 keys directly in the HSMs.

eriktate commented:
> Using SHA-256 for the RSA 3072 OAEP is fine, as SHA-256 is at the 128 bit security level.

Looking through documentation again I might be incorrect about RSA 2048 being a limitation. Let me do some experimentation and I'll update the RFD if making the switch is possible.

> I don't understand why you use an asymmetric algorithm for the key wrapping. As key wrapping and unwrapping will be done on a trusted device (e.g., a network HSM), there is no benefit using an asymmetric key - the key will never leave the device.

This is totally reasonable and I think the solution you proposed below would be preferred if we had guarantees around all auth servers having access to the same keys. That's currently not true, though, since the key store configuration is per auth server rather than per cluster. You could have auth servers in the same cluster that use different HSMs or even a combination of an HSM and a cloud KMS. Wrapping with asymmetric keys enables sharing of recording encryption keys without needing access to the same key store.

This is actually even trickier for HSM backends because Teleport includes the auth server's host UUID in the key name, which essentially ensures that keys can't be shared across auth servers even if they have access to the same networked HSMs.

> Keep in mind that you need to do key rotation both on the data encryption key (currently X25519) and the key encryption key (currently RSA). So in my opinion, your best bet for doing this is:
> 
> * Every auth server has access to all wrapped keys, e.g. through a shared database.
> * Every auth server has access to the trusted key decryption device(s)/server(s) - e.g. networked HSM(s)
> * When an auth server requires access to a recording, it searches for the correct wrapped decryption key in the shared database, based on customer configured key rotation and data retention settings. E.g., if key rotation happens yearly and data retention is for 3 years, then you have 4 data encryption keys (X25519), and 4 key encryption keys (AES - preferred, or RSA as per the current design). When a recording from the previous year needs to be replayed, the auth server requests the correct wrapped private key from the DB and then connects to the key decryption device to obtain the decrypted wrapped key required for playback. The decryption device uses the key encryption key for that year to decrypt the data encryption key, and then returns the data encryption key.
>   However, this step has potential for optimisation. Looking at the `age` file format, it already uses the concept of a data encryption key (ChaCha20-Poly1305) encrypted with a key encryption key (X25519). So instead of wrapping the X25519 key with another key encryption key, it seems to be more efficient to host the X25519 keys directly in the trusted key decryption device(s). All auth servers need to have access to the trusted key decryption device(s) anyway. Then you can directly decrypt the ChaCha20-Poly1305 data decryption key on the trusted key decryption device and send that back. This saves you one level of indirection - one decryption per request. You get the additional benefit that the ChaCha20-Poly1305 keys are unique per recording file, so even if one of them leaks from the auth servers, there is very limited damage done. Whereas, if one of the X25519 private keys leaks, then all recordings from that given year are accessible.
>   It has the further benefit that you do not need to maintain a database with wrapped X25519 keys that is accessible for all clients. The recordings contain all the information you need already.
> * You can still distribute the X25519 public keys to all the recording servers, as they can easily be exported from the trusted key decryption device.

We're still planning on building out rotations for key encryption keys, it's just been omitted from this RFD so we can spend some more time making sure we handle all of the known edge cases properly. Like I mentioned above, your proposed solution would be dramatically simpler but it's not guaranteed that all auth servers have access to the same trusted device or the same keys.

There are definitely optimizations that could be made if we targeted a more constrained usage, but would the general implementation proposed in the RFD be insufficient for your use case?

eriktate commented on rfd/0127-encrypted-session-recordings.md:
````diff
@@ -0,0 +1,843 @@
+
+```yaml
+# session_recording_config.yml
+kind: session_recording_config
+version: v2
+spec:
+  encryption:
+    # whether or not encryption should be enabled
+    enabled: true
+    # whether or not Teleport should manage the keys or simply consume them
+    key_manager: 'keystore|teleport'
+    # key labels Teleport will use to find active encryption keys when mode is
+    # set to "keystore-managed"
+    active_key_labels: []
+    # key labels Teleport will use to find rotated encryption keys when mode is
+    # set to "keystore-managed"
+    rotated_key_labels: []
````
Yeah I think this is clearer. It also makes it easier to add config options for manual management in the future :+1:

What do you think about calling them `labels`? That's sort of specific to PKCS#11, but if you wanted to use this with an AWS KMS backend I think it would have to be a list of IDs, ARNs, or aliases. Is reusing the same config confusing? We could have a list of tagged values:
```yml
active_keys:
  - type: aws_id
     key: mrk-1234abcd12ab34cd56ef1234567890ab
  - type: gcp_id
    key: projects/proj-1/locations/global/keyRings/ring-1/cryptoKeys/teleport_recording/cryptoKeyVersions/1
  - type: pkcs11_label
    key: teleport_recording
```
Or maybe even have an object that supports all of the possible field names:
```yml
active_keys:
  - aws_id: mrk-1234abcd12ab34cd56ef1234567890ab
  - gcp_id: projects/proj-1/locations/global/keyRings/ring-1/cryptoKeys/teleport_recording/cryptoKeyVersions/1
  - pkcs11_label: teleport_recording
```
Although this would require validating against weird configurations where a single key has an `aws_key` and a `pkcs11_label` configured, so maybe the tagged version makes more sense.
 
Some sort of differentiation would also make it easier to migrate to a new keystore in manual mode because you could explicitly define keys for multiple backends. Right now we'd either have to inspect the label value to figure out which backend it belongs to or try all labels and ignore failures when using something like `teleport_recording` as an AWS ID.

eriktate commented on rfd/0127-encrypted-session-recordings.md:
```diff
@@ -0,0 +1,843 @@
+see that their key has been fulfilled, and finish the import process using
+their configured keystore.
+
+Some keystores support this sort of wrapped key exchange without ever exposing
```
This was actually duplicating most of the key sharing section a couple of paragraphs below. It also isn't relevant to the proposed direction of avoiding key exchange altogether, so it doesn't really make sense to include it here. I went ahead and removed this and added some of the missing parts to the key sharing section below.

I thought about removing key sharing altogether, but I think it's probably worth keeping some of it documented just in case we need to implement something similar in the future.

eriktate commented on rfd/0127-encrypted-session-recordings.md:
```diff
@@ -0,0 +1,843 @@
+configured labels. Note that the "keys" cached for KMS and HSM backends are
+just references to real keys stored in the keystore.
+
+### Key Rotation
```
Yeah this is definitely something we're thinking about. Key rotation as it's currently designed is meant to limit the blast radius of a given breached key. The timeline for when those rotated keys are fully deleted ultimately comes down to the retention policy of the recording data itself.

eriktate commented on rfd/0127-encrypted-session-recordings.md:
```diff
@@ -0,0 +1,836 @@
+// RecordingEncryptionKey represents an asymmetric encryption key pair usable
+// during recording encryption and decryption.
+message RecordingEncryptionKey {
+  // The asymmetric keypair used to wrap the age file key. Expected to be RSA.
```
Line 47 was actually describing what's supported by `age` out of the box, but you're right that the RFD proposes only RSA 4096. I can update that paragraph to highlight the distinction :+1:

eriktate commented on rfd/0127-encrypted-session-recordings.md:
```diff
@@ -0,0 +1,836 @@
+
+- AWS KMS supports setting an algorithm of `RSAES_OAEP_SHA_256` when using the
+  `Decrypt` action of the KMS API.
+- GCP CKM supports generating keys of type `RSAES_OAEP_2048_SHA256` which can
```
It does, I missed updating this line when switching everything from 2048. I'll fix that :+1:

eriktate commented on rfd/0127-encrypted-session-recordings.md:
```diff
@@ -0,0 +1,836 @@
+shared, such as a software key or an AWS KMS key that all auth servers have
+access to. This RFD proposes that recording encryption should assume that
+keystores and keys are available to all auth servers in a cluster. This would
+require either adjusting HSM key IDs to remove the `host_uuid` or ignoring the
```
This is still true for sign/verify use cases but not for encryption cases. Manual key management won't have any assumptions about labels other than what's provided in the configuration

eriktate commented on rfd/0127-encrypted-session-recordings.md:
```diff
@@ -0,0 +1,836 @@
+- Generate an asymmetric data key, export the encrypted
+  private key, and save it for future sharing.
+- Import the generated asymmetric key back into KMS. This requires reencrypting
+  in software and exposes the private key to the Teleport process.
```
> Or is the idea that the software keys are only exposed for a short time and compromise is less likely during that time?

Yeah that would be the idea, but we aren't actually implementing this right now. I included it as documentation for a method that might be possible if we needed this type of key sharing in the future, but it could also be omitted from the RFD. I'll see if I can make it more explicit that this is just documenting a suggestion rather than describing what we're going to deliver for this feature

eriktate commented on rfd/0127-encrypted-session-recordings.md:
```diff
@@ -0,0 +1,836 @@
+`session_recording_config.status.encryption_keys`. Each unique public key will
+result in a custom `Recipient` implementation responsible for wrapping the file
+key using `OAEP-SHA256`. They will include the fingerprinted public key in the
+stanza to make lookup during replay easier. The `Identity` implementation will
```
I wasn't planning on accepting or exposing the fingerprint in the configuration. Instead a fingerprint will be generated for the public key after it's fetched from the keystore. The fingerprint itself will be a bas64 encoded sha256 hash of the public key in PKIX ASN.1 DER form.

eriktate commented on rfd/0127-encrypted-session-recordings.md:
````diff
@@ -0,0 +1,634 @@
+```bash
+tsh play 49608fad-7fe3-44a7-b3b5-fab0e0bd34d1.tar
+
+"Replaying encrypted recording files is not supported by tsh, try replaying the with the session ID instead"
````
Since the auth service is the only place where Teleport integrates with the keystore, `tsh` won't have access to unwrap/decrypt the `age` file key. It would require round-tripping the file to the auth service first, at which point you should be able to stream the session by ID anyway.

gl-mc replied:
Files could be renamed from their orginial ID, especially if they are 'valuable', e.g. in court cases or similar. Then it's easier to be able to replay the whole file than having to find the session ID again.

eriktate commented on rfd/0127-encrypted-session-recordings.md:
```diff
@@ -0,0 +1,634 @@
+When using KMS keystores, auth servers may share access to the same key. In
+that case, they will also share the same wrapped key which can be identified by
+the key ID. For HSMs integrated over PKCS#11, the auth server's host UUID is
+attached to the key ID and will be used to determine access.
```
The UUID is stored in the label of the HSM key, yeah. Teleport has a keystore-independent concept of a key ID that includes the host UUID to help identify specific keys and prevent access from other auth servers. For the encryption use case we would relax this and allow for shared keys on a networked HSM.

eriktate commented on rfd/0127-encrypted-session-recordings.md:
```diff
@@ -0,0 +1,634 @@
+// KeySet contains the list of active and rotated WrappedKeys for a
+// given usage.
+message KeySet {
+  // AciveKeys is a list of active, wrapped X25519 private keys. There should
```
Our proto definitions used to be pascal case but we've switched over to snake case for newer protos. I originally had the wrong casing and switched everything. Including the pascal case in the comments was also to comply with godoc requirements in the generated `*.pb.go` files, but I think we've dropped that as a requirement as well

eriktate commented on rfd/0127-encrypted-session-recordings.md:
```diff
@@ -0,0 +1,634 @@
+// WrappedKey wraps a PrivateKey using an asymmetric keypair.
+message WrappedKey {
+  // RecordingEncryptionPair is the asymmetric keypair used to wrap the
+  // private key. Expected to be RSA.
```
I think we've discussed this elsewhere, but RSA with OAEP using SHA256 was chosen because it's supported across all of our keystores (AWS KMS, GCP CKM, and HSMs). We've also dropped the intermediate asymmetric key and key sharing from the scope so the symmetric `age` file keys should be the only keys being wrapped now.

espadolini commented on rfd/0127-encrypted-session-recordings.md:
```diff
@@ -0,0 +1,849 @@
+encryption format for session recordings. It was chosen for its provenance,
+simplicity, and focus on strong cryptography defaults without requiring
+customization. The formal spec can be found
+[here](https://age-encryption.org/v1). The officially supported key wrapping
+algorithms for `age` are limited to X25519, Ed25519, and RSA. Support for other
```
When you say ed25519 and RSA are you referring to the SSH key support in age?

eriktate replied:
Yep exactly that

espadolini commented on rfd/0127-encrypted-session-recordings.md:
```diff
@@ -0,0 +1,849 @@
+compatible with FIPS, which means configuring encrypted sessions while in FIPS
+mode will result in failed startup.
+
+In order to support key unwrapping directly within HSM and KMS systems, this
+RFD proposes a custom `age` plugin that uses RSA 4096 keypairs in combination
+with the decryption APIs exposed by each keystore backend. The choice of RSA is
```
If age supports RSA natively why do we need a plugin?

espadolini commented on rfd/0127-encrypted-session-recordings.md:
```diff
@@ -0,0 +1,849 @@
+mode will result in failed startup.
+
+In order to support key unwrapping directly within HSM and KMS systems, this
+RFD proposes a custom `age` plugin that uses RSA 4096 keypairs in combination
```
Are we going with RSA-4096 for some specific security level that the more common RSA-2048 does not provide? If so, please mention it.

eriktate replied:
We are, 128-bit is the minimum. I've updated the RFD to mention that :+1:

espadolini commented on rfd/0127-encrypted-session-recordings.md:
```diff
@@ -0,0 +1,849 @@
+// EncryptionKeyPair is a keypair used for encrypting and decrypting data.
+message EncryptionKeyPair {
+  // public_key is the public encryption key.
+  bytes public_key = 1 [(gogoproto.jsontag) = "public_key"];
```
Don't use gogoproto features for anything new.

espadolini commented on rfd/0127-encrypted-session-recordings.md:
```diff
@@ -0,0 +1,849 @@
+
+// KeyState represents that possible states a RecordingEncryptionKey can be in.
+enum KeyState {
+  // Default KeyState
+  KEY_STATE_UNSPECIFIED = 0;
```
Do not ascribe any meaning to the zero value of an enum, it's not a default, it's a missing value and if it's necessary to specify it and it's set to zero then the message should be treated as not valid (as if the value was set to some other value that's completely unknown).

espadolini commented on rfd/0127-encrypted-session-recordings.md:
```diff
@@ -0,0 +1,849 @@
+// api/proto/teleport/legacy/types/types.proto
+
+// EncryptionKeyPair is a keypair used for encrypting and decrypting data.
+message EncryptionKeyPair {
```
EncryptionKeyPair doesn't need to be in api/proto/teleport/legacy/types/types.proto, right?

espadolini commented on rfd/0127-encrypted-session-recordings.md:
```diff
@@ -0,0 +1,849 @@
+  uint32 hash = 4 [(gogoproto.jsontag) = "hash"];
+}
+
+// AgeEncryptionKey is a PEM encoded RSA4096 public key used for encrypting
+// session recordings.
```
Don't use PEM especially if it's stored in a `bytes` field, store the spki in DER form instead (which could likely not need anything more to specify that it's a RSA public key rather than something else).

espadolini commented on rfd/0127-encrypted-session-recordings.md:
```diff
@@ -0,0 +1,849 @@
+For keys not managed by Teleport, key sharing is left to the administrator to
+facilitate.
+
+In order for all auth servers in a cluster to replay all recordings, they will
+need access to the same keys. There are common cases where keys are easily
+shared, such as a software key or an AWS KMS key that all auth servers have
+access to. This RFD proposes that recording encryption should assume shared
+access to the same keys and keystores from all auth servers in a cluster. For
+networked HSMs, this will require replacing the `host_uuid` label applied to
+keys with a shared label signalling recording encryption.
```
If all auth servers have access to the REKs why do we need to keep track of unaccessible keys and individual key rotations and multiple active keys at once?

eriktate replied:
Keeping track is just a failsafe in case that assumption breaks down because of a bad auth server configuration. Multiple active keys should only be possible during key rotations before the completion or rollback RPC is called, otherwise there should just be a single active key.
